beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+/// @param grid_cell_data The grid cell data map.
+/// @param spatial_hash_function_ The hash object instance.
+/// @param neighbors Range containing the neighbors of a cell.
+/// @param weight_cap The maximum weight value to be assigned to each cell.
+template <class GridCellDataMapType, class Hasher, class Neighbors>
+void map_cells_to_clusters(
+    GridCellDataMapType&& grid_cell_data,
+    Hasher&& spatial_hash_function_,
+    Neighbors&& neighbors,
+    double weight_cap) {
+  auto grid_cell_queue = populate_priority_queue(grid_cell_data);
+
+  std::size_t next_cluster_id = 0;
+
+  while (!grid_cell_queue.empty()) {
+    const auto grid_cell_hash = grid_cell_queue.top().hash;
+    grid_cell_queue.pop();
+
+    // any hash that comes out of the queue is known to exist in the cell data map
+    auto& grid_cell = grid_cell_data[grid_cell_hash];
+    const auto& grid_cell_weight = grid_cell.weight;
+    const auto& representative_pose_in_world = grid_cell.representative_pose_in_world;
+
+    // if there's no cluster id assigned to the cell, assign it a new one
+    if (!grid_cell.cluster_id.has_value()) {
+      grid_cell.cluster_id = next_cluster_id++;
+    }
+
+    // process the neighbors of the cell; if they don't have a cluster id already assigned
+    // then assign them one and add them to the queue with an inflated priority
+    // to ensure they get processed ASAP before moving on to other local peaks.
+    // Notice that with this algorithm each cell will go through the priority queue at most twice.
+
+    const auto get_neighbor_hash = [&](const auto& neighbor_pose_in_representative) {
+      return spatial_hash_function_(representative_pose_in_world * neighbor_pose_in_representative);
+    };
+
+    const auto filter_invalid_neighbors = [&](const auto& neighbor_hash) {
+      auto it = grid_cell_data.find(neighbor_hash);
+      return (
+          (it != grid_cell_data.end()) &&            // is within the map
+          (!it->second.cluster_id.has_value()) &&    // AND not yet mapped to a cluster
+          (it->second.weight <= grid_cell_weight));  // AND has lower weight than the current cell
+    };
+
+    auto valid_neighbor_hashes_view =                     //
+        neighbors |                                       //
+        ranges::views::transform(get_neighbor_hash) |     //
+        ranges::views::filter(filter_invalid_neighbors);  //
+
+    for (const auto& neighbor_hash : valid_neighbor_hashes_view) {
+      auto& neighbor = grid_cell_data[neighbor_hash];
+      neighbor.cluster_id = grid_cell.cluster_id;
+      const auto inflated_priority =
+          weight_cap + neighbor.weight;  // since weights are capped at weight_cap, this gives us a value that is
+                                         // guaranteed to be higher than any other weight from a local maximum.
+      grid_cell_queue.push({inflated_priority, neighbor_hash});  // reintroduce with inflated priority
```

Comment:
@glpuga meta: to flood assign cluster ids from a given local maximum I presume. Is there a textbook / article reference for this algorithm? How does it fare against other clustering algorithms?

Reply:
> to flood assign cluster ids from a given local maximum I presume.
Indeed, with additional provisions to avoid cluster fragmentation due to pdf estimation noise.

> Is there a textbook / article reference for this algorithm?
Nope.

> How does it fare against other clustering algorithms?
God knows. Doing such a comparison would be a paper on its own we may someday write.